bitnami/clickhouse/templates/service.yaml

```diff
@@ -3,150 +3,186 @@ Copyright Broadcom, Inc. All Rights Reserved.
+  {{- if and (eq .context.Values.service.type "LoadBalancer") (not (empty .context.Values.service.loadBalancerSourceRanges)) }}
+  loadBalancerSourceRanges: {{- toYaml .context.Values.service.loadBalancerSourceRanges | nindent 4 }}
   {{- end }}
-  {{- if and (eq .Values.service.type "LoadBalancer") (not (empty .Values.service.loadBalancerSourceRanges)) }}
-  loadBalancerSourceRanges: {{- toYaml .Values.service.loadBalancerSourceRanges | nindent 4 }}
+  {{- if eq .context.Values.service.type "LoadBalancer" }}
+  {{- if and .context.Values.service.perReplicaAccess (not (empty .context.Values.service.loadBalancerIPs)) }}
+  {{- $posOffset := mul (int .shard) (.context.Values.replicaCount | int) }}
+  {{- $position := add $posOffset (int .replica) }}
+  loadBalancerIP: {{ index .context.Values.service.loadBalancerIPs $position }}
+  {{- else if not (empty .context.Values.service.loadBalancerIP) }}
+  loadBalancerIP: {{ .context.Values.service.loadBalancerIP }}
   {{- end }}
-  {{- if and (eq .Values.service.type "LoadBalancer") (not (empty .Values.service.loadBalancerIP)) }}
-  loadBalancerIP: {{ .Values.service.loadBalancerIP }}
   {{- end }}
   ports:
     - name: http
       targetPort: http
-      port: {{ .Values.service.ports.http }}
+      port: {{ .context.Values.service.ports.http }}
       protocol: TCP
-      {{- if and (or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer")) (not (empty .Values.service.nodePorts.http)) }}
-      nodePort: {{ .Values.service.nodePorts.http }}
-      {{- else if eq .Values.service.type "ClusterIP" }}
+      {{- if and .context.Values.service.perReplicaAccess (not (empty .context.Values.service.nodePorts.http)) }}
+      {{- $posOffset := mul (int .shard) (.context.Values.replicaCount | int) }}
+      {{- $position := add $posOffset (int .replica) }}
+      nodePort: {{ index .context.Values.service.nodePorts.http $position }}
+      {{- else if and (or (eq .context.Values.service.type "NodePort") (eq .context.Values.service.type "LoadBalancer")) (not (empty .context.Values.service.nodePorts.http)) }}
+      nodePort: {{ .context.Values.service.nodePorts.http }}
+      {{- else if eq .context.Values.service.type "ClusterIP" }}
       nodePort: null
       {{- end }}
-    {{- if .Values.tls.enabled }}
+    {{- if .context.Values.tls.enabled }}
     - name: https
       targetPort: https
-      port: {{ .Values.service.ports.https }}
+      port: {{ .context.Values.service.ports.https }}
       protocol: TCP
-      {{- if and (or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer")) (not (empty .Values.service.nodePorts.https)) }}
-      nodePort: {{ .Values.service.nodePorts.https }}
-      {{- else if eq .Values.service.type "ClusterIP" }}
+      {{- if and .context.Values.service.perReplicaAccess (not (empty .context.Values.service.nodePorts.https)) }}
+      {{- $posOffset := mul (int .shard) (.context.Values.replicaCount | int) }}
```

Comment:
I would add a {{- /* comment */}} explaining a bit what we're doing with the offsets in the nodePort section, as it is a bit difficult to understand at first right

Reply:
done at https://github.com/bitnami/charts/pull/32958/commits/786dc3e86a09575f9fb564f1362f6c4b56c4e837